### ISSUE TITLE
Names for expanded tasks

### ISSUE THREAD

**Issue by MatrixManAtYrService** [apache/airflow#23020]
### Description

Airflow currently exposes `map_index` to the user as a way of distinguishing between tasks in an expansion.  The index is unlikely to be meaningful to the user.  They probably have their own label for this action.  I'm requesting that we allow them to add that label.

To see the problem, consider a dag that sends email to a list of users which is generated at runtime:

```python3
with DAG(...) as dag:

    @dag.task
    def get_account_status():
        return [
            {
                "NAME": "Wintermute",
                "EMAIL": "wintermute@tessier-ashpool.com",
                "STATUS": "active",
            },
            {
                "NAME": "Hojo",
                "EMAIL": "ops@research.shinra.com",
                "STATUS": "delinquent",
            },
        ]

    BashOperator.partial(
        task_id="send_email",
        bash_command=dedent(
            """
            cat <<- EOF | tee | mailx -s "your account" $EMAIL
            Dear $NAME,
                Your account status is $STATUS.
            EOF
            """
        ),
    ).expand(env=get_account_status())
```

Notice that in the grid view, it's not obvious which task goes with which user:

<img width="1106" alt="Screen Shot 2022-04-14 at 8 56 09 AM" src="https://user-images.githubusercontent.com/5834582/163418431-3180a29d-b9c0-4bbc-9a80-05ad5e4f34e7.png">

### Use case/motivation

I'd like to be able to explicitly assign a name to each expanded task, that way I can later go look at the right one.  I would like this name to be used (when available) anywhere that the user interacts with the expanded task.

In cases where the user provides no names, perhaps we can generate some.  For instance, this expansion generates four instances.

```python3
BashOperator.partial(task_id="greet").expand(
    bash_command=["echo hello $USER", "echo goodbye $USER"],
    env=[{"USER": "foo"}, {"USER": "bar"}],
)
```

The friendliest way would be to use the requested feature name each task:
- `hi_foo`
- `hi_bar`
- `bye_foo`
- `bye_bar`

As it is, the user will see:

- `1`
- `2`
- `3`
- `4`

But if the user doesn't give names, maybe we should generate some names for them:
- `bash_command_1_env_1`
- `bash_command_1_env_2`
- `bash_command_2_env_1`
- `bash_command_2_env_2`

I don't know.  I'm creating this issue so we have a place to discuss it.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by MatrixManAtYrService** [comment]
When I think about how the API might look, I come up with this:

```python
BashOperator.partial(task_id="say").expand_dicts(
    bash_command={
        "generic": "echo hello",
        "cow": "cowsay $MESSAGE",
        "pikachu": "pokemonsay -p Pikachu $MESSAGE",
    },
    env={
        "pikachu": {"MESSAGE": "pika"},
        "cow": {"MESSAGE": "moo"}
    },
)
```

This would create three tasks, and the "generic" one wouldn't define any env vars.

**Comment by bbovenzi** [comment]
+1 I'm super happy to do this in the UI as soon as it can access that data.

**Comment by uranusjr** [comment]
We could use a heuristic similar to pytest. I’m not sure what their exact logic is, but it seems to be something like if the mapped value is simple (int, float, or a short-ish str), use `str()` to convert it directly, otherwise use something like `arg1` (we could use `[0]`, `[1]` etc.).

Some things I can think of now:

1. Calculating the label may be expensive since this requires the view to re-pull upstream XCom. We may need to do this eagerly when the task instance is being executed, which means adding a new field on TaskInstance to store this for later use.
2. The logic becomes difficult when we eventually add functional tools to task mapping e.g. `produce`, `zip` etc. 
3. Pytest allows using `id` to supply custom IDs when the auto-generated label is not good enough (but you don’t want the non-descriptive `arg1`). We can’t do this since the indexes are dynamically calculated, but maybe an argument on BaseOperator to let the user provide a string template for this would be useful.

**Comment by potiuk** [comment]
3. How about (on top of auto-generated id)  add a function to return the task_id(). If we calculate it eagerly at task execution time, there are no security concerns. and it will be far more versatile than template-based id.

**Comment by MatrixManAtYrService** [comment]
re: `zip`, I was thinking that this would make that easier.  If you want to zip together two lists of length _n_ then you can just use a dict comprehension that gives each item one of _n_ keys--items with the same key get zipped together.  I guess it wouldn't really extend to stuff like `accumulate` though.

I didn't include user-defined functions in my proposal because I was worried that:
- If something goes wrong in that function, how do you communicate to them which one had the problem?
- User has to wonder what airflow will do if they return the same key twice--with dicts the collision is handled when the user is building the object.
- Other scary runtime concerns?

It's good to know that maybe my third concern isn't so scary after all.

**Comment by uranusjr** [comment]
Exposing a function might not be a good idea, since the function needs to be executed by the scheduler.

**Comment by potiuk** [comment]
I thought it might be run at execution time and update task instance then. It won't work for queued state but maybe that's not really needed. 

If we allow Jinja template (one of the options) to be run in scheduler it's also a bit dangerous 

Clear demarcation that you update it at execution time should be fine. This is what for example GitHub Actions do. They use the expression language and display the expression until the task start to be executed

**Comment by MatrixManAtYrService** [comment]
Could it go in a separate task that goes upstream of the expansion? That gives us a way to communicate errors back to the user.

I'm imagining such structures appearing in the UI as task groups that you can't expand until they've run.

**Comment by uranusjr** [comment]
It’s possible, but I think separating the logic into task would be much more complicated to implement, since we don’t really have a good way to communiate arbitrary information between tasks (except XCom which is probably not the right place for this). Doing it as a part of the task run as Jarek suggested is better. We already have pre-execute hooks, so it can fit around there.

**Comment by potiuk** [comment]
@MatrixManAtYrService  -  maybe you want to take a stab at it and implement it :) ?

**Comment by MatrixManAtYrService** [comment]
I'd like to, but first I'll have to argue with a certain kanban board.  I'll report back with a link to a draft PR if successful.

**Comment by Gollum999** [comment]
For the sake of discussion, how do people feel about letting these names entirely *replace* `map_index`?

My primary motivation is that an integer-based `map_index` can get weird if the list of tasks changes when re-running a DAG.  In my experience, you can end up with missing and/or duplicate tasks due to the mismatched indices.  As a result you are generally forced to re-run all of the mapped task instances (if not the entire DAG Run), even if 99% of the tasks are unchanged.  I can give more concrete examples if anyone is interested.

I'd propose letting the mapped args become (part of?) the primary key for the mapped task.  This would solve OP's goal of improving the UX of mapped tasks (no more abstract indices), but would also allow for some amount of consistency between runs (not just for re-runs, but potentially between separate DAG Runs as well).

Conceptually I think this is doable since the args already must be JSON serializable, but I'm sure the implementation would be more complex than I am imagining.

A couple considerations that come to mind:
1. These IDs ultimately need to be unique.  So something like `map_index` might still be needed to augment the key and resolve duplicates.  I imagine this could be similar to how duplicate task_ids are resolved (`task`, `task__1`, etc.).
2. The values used to generate the list of dynamic tasks could be very large and complex.  So it seems like you would need the ability to override the generated name with a "short name" when args are too complex.

**Comment by potiuk** [comment]
I think you can forget about this.

You've just hit reality train.

Look there: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping - see this part:

"Rather than overloading the task_id argument to `airflow tasks run` (i.e. having a task_id of `run_after_loop[0]`) we will add a new `--mapping-id` argument to `airflow tasks run` -- this value will be a ~JSON-encoded~ an integer specifying the index/position of the mapping."

We have to support MySQL and the problem with MySQL is that index size is limited. VERY limited. Depending on the type of encooding it might be even 720 characters. And task-id + dag_id  + task_index already exceed the limit. And there is no way around it - and this was the main reason (I believe) we had to use integer, even if originally we planned  not even a name but JSON-encoded list of parameters( which was far better for uniqueness - because it was automated).

But this is just what I saw - by observing it being implemented, so I might be wrong on that account

**Comment by potiuk** [comment]
BTW. I hate MySQL and how it limits us in design, and I proposed already few times to remove it, but we came to the conclusion that supporting the community running MySQL installation historically is more important.  This is unfortunately one of the limitations it introduces.

**Comment by uranusjr** [comment]
Even without the MySQL limitation, we still need some sort of index to keep track of the ordering, otherwise there’s no way to know which upstream value maps to which worker.

**Comment by uranusjr** [comment]
It just occurred to me that this is essentially a part of #22073. What we (users) actually want is a more customisable way to identify things (in this instance, a mapped task _instance_), and if we look past the assumption that a mapped task instance is “task_id + map_index”, we simply need a better way for the user to tell “what is this thing” in the Airflow UI. So let’s keep track of that issue instead to make sure whatever solution we come up for it correctly considers map_index.

**Comment by potiuk** [comment]
> It just occurred to me that this is essentially a part of #22073. What we (users) actually want is a more customisable way to identify things (in this instance, a mapped task _instance_), and if we look past the assumption that a mapped task instance is “task_id + map_index”, we simply need a better way for the user to tell “what is this thing” in the Airflow UI. So let’s keep track of that issue instead to make sure whatever solution we come up for it correctly considers map_index.

Fully agree

**Comment by potiuk** [comment]
Closing this one as duplicate of #22073  (and I make a comment on that issue to make sure it includes mapped tasks).

**Comment by chrisluedtke** [comment]
Can this issue please be re-opened? The implementation of PR #32520 to deliver #22073 does not enable users to define unique id's (map index) for mapped task instances.

Maybe the desired behavior could be added to `.expand()` as a reserved `map_index` kwarg?

```python
import pendulum

from airflow.decorators import dag, task

@dag(
    schedule=None,
    start_date=pendulum.datetime(2023, 1, 1, tz="UTC"),
    catchup=False,
    tags=["example"],
)
def demo():
    @task
    def handle_animal(animal: dict):
        print("Run away")

    animals = [{'id': 'lion'}, {'id': 'tiger'}, {'id': 'bear'}]

    handle_animal.expand(animal=animals, map_index=[a['id'] for a in animals])
```

**Comment by potiuk** [comment]
> Can this issue please be re-opened? The implementation of PR https://github.com/apache/airflow/pull/32520 to deliver https://github.com/apache/airflow/issues/22073 does not enable users to define unique id's (map index) for mapped task instances.

Reopening issue like that does not change much. Someone willl have to. implement it. If you have idea how to implement, you can definitely propose PR. You can also make suggestions in the PR opened. We are not "Issue-driven" we are "code driven" - so PRs are more important than issues, especially on new (small-ish) features like that.

**Comment by chrisluedtke** [comment]
> Reopening issue like that does not change much. Someone willl have to implement it.

Yes, the purpose of an issue is not to implement a feature. The purpose of an issue is to gather interest and discussion around a topic and to inspire implementation.

> We are not "Issue-driven" we are "code driven" - so PRs are more important than issues, especially on new (small-ish) features like that.

The importance of PRs versus issues is not relevant. The original feature request is valid and should be kept open until the feature is delivered.

**Comment by potiuk** [comment]
@chrisluedtke . sure no problem. It does not change a thing to open an issue that has not been discussed for a year. But let me reopen and assign to you, hopefully you can lead it to completion.

**Comment by paulg-mscience** [comment]
It looks like this issue has lost moment but we really need some sort of tag or label attribute that could be assigned per mapped task instance and surfaced in the airflow UI to aide usability, trouble-shooting and monitoring. This tag/label would be in addition to the numeric index and not a replacement for the index.

**Comment by paulg-mscience** [comment]
@kaxil Is this something you can influence and push through?

**Comment by potiuk** [comment]
> It looks like this issue has lost moment but we really need some sort of tag or label attribute that could be assigned per mapped task instance and surfaced in the airflow UI to aide usability, trouble-shooting and monitoring. This tag/label would be in addition to the numeric index and not a replacement for the index.

No. It did not loose moment. Simply no-one picked it up. As mentioned above - the best thing to do when you care about something to be implemented is to ... implement it. Second best is to find someone who can implement it. Possibly sponsor somoene's work.  I thought I explained it above that reopening an issue does not **magically** make it implemented. You need someone who will put their time and effort into it.

**Comment by RNHTTR** [comment]
I'm interested in taking this on (or at least moving it forward). Would it be sufficient to start by:

1. Adding something like `named_mapped_index` to BaseOperator and as a column in the `TaskInstance` model
2. Rendering the named index in the UI in the grid view and TaskInstance view

**Comment by kaxil** [comment]
> I'm interested in taking this on (or at least moving it forward). Would it be sufficient to start by:
> 
> 1. Adding something like `named_mapped_index` to BaseOperator and as a column in the `TaskInstance` model
> 2. Rendering the named index in the UI in the grid view and TaskInstance view

Yes. that's a good start.

@uranusjr can help guide you Ryan if you need any help.

**Comment by damjad** [comment]
@RNHTTR are you working on it?

**Comment by RNHTTR** [comment]
> @RNHTTR are you working on it?

yep! https://github.com/apache/airflow/pull/36797

**Comment by uranusjr** [comment]
Implemented in #36797.

**Comment by pykenny** [comment]
Not trying to reopen the issue (since feature to set displayed name has been shipped). I think the idea of letting users to define their own set of index ID not only hits functional limit on MySQL, but also can lead to the problem similar to this report:

 - https://github.com/apache/airflow/issues/24701

Users can unintentionally generate duplicated IDs during the mapping process, and both overwriting duplicated one (last-write-wins) or appending suffix (like how task IDs are determined) may not fit user expectation.
